src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,395 @@
+            FeaturesResources.ResourceManager,
+            typeof(FeaturesResources));
+
+        private static readonly LocalizableString Description = new LocalizableResourceString(
+            nameof(FeaturesResources.Invalid_format_string),
+            FeaturesResources.ResourceManager,
+            typeof(FeaturesResources));
+
+        private static DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+            DiagnosticID,
+            Title,
+            MessageFormat,
+            DiagnosticCategory.Compiler,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true,
+            description: Description);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+            => ImmutableArray.Create(Rule);
+
+        // this regex is used to remove escaped brackets from
+        // the format string before looking for valid {} pairs
+        private static Regex s_removeEscapedBracketsRegex = new Regex("{{");
+
+        // this regex is used to extract the text between the
+        // brackets and save the contents in a MatchCollection
+        private static Regex s_extractPlaceholdersRegex = new Regex("{(.*?)}");
+
+        private const string NameOfArgsParameter = "args";
+        private const string NameOfFormatStringParameter = "format";
+
+        protected abstract ISyntaxFactsService GetSyntaxFactsService();
+        protected abstract TSyntaxKind GetInvocationExpressionSyntaxKind();
+        protected abstract SyntaxNode GetArgumentExpression(SyntaxNode syntaxNode);
+        protected abstract ITypeSymbol GetArgumentExpressionType(SemanticModel semanticModel, SyntaxNode argsArgument);
+        protected abstract SyntaxNode GetMatchingNamedArgument(SeparatedSyntaxList<SyntaxNode> arguments, string searchArgumentName);
+        protected abstract bool ArgumentExpressionIsStringLiteral(SyntaxNode syntaxNode);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(startContext =>
+            {
+                var formatProviderType = startContext.Compilation.GetTypeByMetadataName(
+                    "System.IFormatProvider");
+                if (formatProviderType == null)
+                {
+                    return;
+                }
+
+                startContext.RegisterSyntaxNodeAction(
+                    c => AnalyzeNode(c, formatProviderType),
+                    GetInvocationExpressionSyntaxKind());
+            });
+        }
+
+        private void AnalyzeNode(SyntaxNodeAnalysisContext context, INamedTypeSymbol formatProviderType)
+        {
+            var optionSet = context.Options.GetDocumentOptionSetAsync(
+                context.Node.SyntaxTree, context.CancellationToken).GetAwaiter().GetResult();
+
+            if (optionSet?.GetOption(
+                ValidateFormatStringOption.ReportInvalidPlaceholdersInStringDotFormatExpression,
```

Comment:
This is what turns the analyzer off, since it's packaged in VS by default right now.

The alternative would be to do as Sam suggested, and move this whole thing to roslyn-analyzers.